### PR TITLE
Move uniform_() to DistributionTemplates(Migrate CPU uniform_ from TH to ATen)

### PR DESCRIPTION
**Pull request by pbelevich**
`uniform_kernel_cpu` is based on https://github.com/pytorch/pytorch/pull/30954

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35585 rand() and uniform_() for complex dtype
* **#35580 Move uniform_() to DistributionTemplates(Migrate CPU uniform_ from TH to ATen)**

Differential Revision: [D20820221](https://our.internmc.facebook.com/intern/diff/D20820221)

### CODE REVIEW

File: aten/src/ATen/native/DistributionTemplates.h [hunk 1]
```diff
@@ -262,4 +260,29 @@ Tensor normal_impl(const Tensor& mean, const Tensor& std, Generator gen) {
   return ret;
 }
 
+// ==================================================== Uniform =======================================================
+
+template<template<typename> class uniform_kernel, typename RNG>
+at::Tensor& uniform_impl_(at::Tensor& self, double from, double to, at::Generator generator) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, self.scalar_type(), "check_uniform_bounds", [&] {
```
Comment (anjali411) [inline aten/src/ATen/native/DistributionTemplates.h:267]: why does the complex handling not show up here? I know you made the changes here: https://github.com/pytorch/pytorch/pull/35585

File: aten/src/ATen/test/cpu_rng_test.cpp [hunk 2]
```diff
@@ -96,13 +102,15 @@ class RNGTest : public ::testing::Test {
...
+      .impl("aten::normal_",                  torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_)))
       .impl("aten::normal.Tensor_float_out",  torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_Tensor_float_out)))
       .impl("aten::normal.float_Tensor_out",  torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_float_Tensor_out)))
       .impl("aten::normal.Tensor_Tensor_out", torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_Tensor_Tensor_out)))
       .impl("aten::normal.Tensor_float",      torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_Tensor_float)))
       .impl("aten::normal.float_Tensor",      torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_float_Tensor)))
       .impl("aten::normal.Tensor_Tensor",     torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(normal_Tensor_Tensor)))
+      // Uniform
+      .impl("aten::uniform_",     torch::dispatch(DispatchKey::CustomRNGKeyId, CppFunction::makeUnboxedOnly(uniform_)))
```
Comment (anjali411) [inline aten/src/ATen/test/cpu_rng_test.cpp:113]: what does this line do?

Comment (pbelevich) [reply aten/src/ATen/test/cpu_rng_test.cpp:113]: it registers custom `uniform_ ` with dispatch key `CustomRNGKeyId ` to "aten::uniform_" schema

File: aten/src/ATen/native/DistributionTemplates.h (hunk 1 above)
Comment (pbelevich) [reply aten/src/ATen/native/DistributionTemplates.h:267]: why should it be here?

**Review by anjali411 — APPROVED** [review]
LGTM :)

**Pull request closed without merging**